src/libstore/build/local-derivation-goal.cc

```diff
@@ -1672,6 +1672,12 @@ void setupSeccomp()
         if (seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), SCMP_SYS(fchmodat), 1,
                 SCMP_A2(SCMP_CMP_MASKED_EQ, (scmp_datum_t) perm, (scmp_datum_t) perm)) != 0)
             throw SysError("unable to add seccomp rule");
+
+#if defined(__SNR_fchmodat2)
```

Comment:
when is this not defined? which package is necessary to build with an old version of, to produce a nix that has inconsistent behaviour on newer kernels? 

i forget where these constants come from but i don't like how this makes runtime behaviour depend on building with new enough headers.

Reply:
`__SNR_fchmodat2` is defined in libseccomp and points to `__NR_fchmodat2` from glibc.

The problem here is that we need glibc 2.39 (which means nixos-unstable) to get the correct syscall number.
We may want to update the flake in this project to actually build against that though.

So I see the following options:
* Updating the flake here. Has the downside that 23.11 still has the issue.
* Inject a newer `<sys/syscall.h>` into the build (and also a patched libseccomp header) and I'm not sure if we want that (_looking at @NixOS/nix-team who has to maintain that later on_).
* The syscall number is 452 except on mips, alpha and x32:

```
❯ ~/Projects/glibc release/2.39/master → rg --pcre2 'fchmodat2 (?!452)' **/arch-syscall.h 
sysdeps/unix/sysv/linux/x86_64/x32/arch-syscall.h
58:#define __NR_fchmodat2 1073742276

sysdeps/unix/sysv/linux/mips/mips64/n32/arch-syscall.h
67:#define __NR_fchmodat2 6452

sysdeps/unix/sysv/linux/mips/mips64/n64/arch-syscall.h
62:#define __NR_fchmodat2 5452

sysdeps/unix/sysv/linux/mips/mips32/arch-syscall.h
70:#define __NR_fchmodat2 4452

sysdeps/unix/sysv/linux/alpha/arch-syscall.h
59:#define __NR_fchmodat2 562
```

  x32 seems to be something different than i686, so none of the architectures here seem supported by Nix right now. As a temporary measure we could hardcode the syscall number until the NixOS ecosystem has moved to glibc 2.39.

I don't really like any of these. Opinions? Or better ideas?

Reply:
my preference would be a shim header that defines it if not yet defined, since it's not like it's going to change anyway. and then put the bad system detection code in that header. least build system shenanigans, and puts things that don't matter in a harmless location.

Reply:
I have a patch now locally that works on my x86_64-linux. Will take a second look tomorrow after having some sleep and push it then.

Reply:
Pushed a fix for this.